### DISCUSSION TITLE
Media "watch folder" on server as an alternative to client upload

### DISCUSSION THREAD

**Discussion by gpatkinson** [mediacms-io/mediacms#310] [General]
I have been experimenting with MediaCMS and really like the software. Thank you for all your hard work on this project.

As far as I can tell, the only way to add media is by uploading from a client. This works, but in my case, I have other server-side software pulling video files. So what would really be helpful is being able to assign a "watch folder" where all files within this folder are processed and auto-added to the database, preferably without needing to duplicate the files.

If this type of feature (or a work-around) exists currently, I would appreciate a quick note about it in the docs.

**Comment by mgogoulos** [comment]
Hello, thanks for your kind words. Besides the uploader, there's also an API that can be used to send new media. Plus there is also a command line client that utilizes this API but is on a separate branch/PR. 

What you describe seems like a  specialized scenario but I think you can easily implement a solution by creating a script that will monitor the folder and POST new media on the 'Add new media' URL (See what it expects on https://demo.mediacms.io/swagger/). Hope this helps

**Reply by tabilzad** [reply]
@mgogoulos any way to by pass transferring files through the network. Is there a `.py` script I could run to provision new files manually added to `media_files` dir and seed the db.

**Reply by tabilzad** [reply]
I ended up creating another endpoint to trigger local file scan and auto create media from those files.

**Reply by mgogoulos** [reply]
Glad this works for you!

**Reply by srvprivate** [reply]
@tabilzad @mgogoulos Can either of you shed more light on this situation? I'm a video producer -- got this up and working via docker compose, but my issue is that all my files live on the server itself. I do not and refuse however to upload duplicate media when it's already sitting in its home. Is there a way to get these directories talking to each other for a 'scan'? API of some sort? I'm novice, but know enough to be efficient. Any help would be much appreciated. This is by far the best video hosting platform I've run across. kudos on such great work. I am ditching vimeo and youtube because of this. That means as a professional, I'm taking this CMS very seriously!

**Reply by mgogoulos** [reply]
Video files *have* to be properly imported by MediaCMS because it performs a number of actions (create preview gif file, poster/thumbnail of different dimensions, encodings for different dimensions/bitrates, HLS files for adaptive streaming). So even if there was a page where you said "my directory is x/y/z, take videos from this and bring them to the system" which is what I am assuming others here are doing, it would still have to get these videos and bring it to media_files/ directory that it is using to server original videos and produced files.

**Reply by srvprivate** [reply]
Yes, you can look at this more like a MaM too. I have used one called Flow by EditShare where they allow mounting of SMB shares (et al) and will scan the directory and import into the database, making proxies, etc. That is all done in python, so I know it's possible!

**Reply by srvprivate** [reply]
Any new news on this? a year later?

**Reply by thomisus** [reply]
> I ended up creating another endpoint to trigger local file scan and auto create media from those files.

Could you share your solution?

**Reply by assassinliujie** [reply]
Actually, I think what many people want is to use the project's UI to index and watch videos from their own existing files, rather than reprocessing existing videos and redistributing them. Is it possible to distribute the original files directly?

**Comment by ttkrpink** [comment]
I am in the same situation where the media files are uploaded to a folder by another program. So, the suggested method will be to upload the files to MediaCMS again using the APIs so MediaCMS can process the files. 

Is that correct?

**Comment by thomisus** [comment]
Same situation here. Media/Videos are on an external server/share that I'd like to not touch because is ready to be edited or archived but I'd like to index and categorize it. 
What would be great is a way to recursively scan a "watch folder" ( or trigger with inotify ) then import metadata and create low res previews and assign it to a default user, leaving original files in the starting location.

**Comment by srvprivate** [comment]
I think we need to formally open a case on this. In 2025, re-uploading is not going to work for many people who have a share or index of files sitting somewhere. Many people have files sitting in numerous places. There is definitely a way to write the code to index files sitting elsewhere. I've used plenty of programs like Flow by EditShare that does an indexing scan, and I know this is possible. Many people would thank you for this! I have not used MediaCMS in a year for this issue alone, and would use it and promote others to if scanning were an option.

**Reply by davidcmsmith** [reply]
I'm in a situation where I have a bunch of pre transcoded files that I would like to bring to this platform, however it seems that I would need to bring the original files to mediacms again in order to index and transcode again.

**Reply by crocodisle** [reply]
Yes, this would be totally awesome! MediaCMS is 90% there for my use case, but it's this sort of feature alone that's keeping me from adopting it. I don't need transcoding or file copying, I just need to point MediaCMS at a directory and have it index/reference those files directly. 

Looks like this is something that's been talked about for a few years at this point, so it continues to be a need. Currently, I can't find any video player that does this that also has an option to embed videos.
#180

**Comment by Clear2437** [comment]
Would be nice to see some progress on this eventually. The ticket is open for 4 years already, I would think this should be quite high in priority as it seems like a basic core feature for a video platform.
Appreciate your work!

**Reply by srvprivate** [reply]
I don't use MediaCMS all the time right now because of this. What can we do to triage something?

I want to use it, but manually uploading is a little dated. There are ways to scan and import with metadata and transcoding. You're not far off. Is it beyond your time/skill at the moment?

**Comment by srvprivate** [comment]
Adding this as an official need for documentation. @mgogoulos I'm going to embark on this API stuff, but could use some help. I really want to see this take off having automatic POST. Like for example, I use TubeArchivist for all my Youtube videos, think about this — auto archiving youtube channels is brilliant, but tying that into MediaCMS is even more brilliant. I chose MediaCMS over Peertube. 

Secondly, MeTube (a derivative of yt-dlp) is excellent for downloading say, Facebook live videos. Think about the flow here, I can automate almost everything, and it ends up in my media library just with a click of a button, and if I want to share something I can give a link! That's my goal here. What is the best place to start on getting this to work?

**Reply by mrjoshuap** [reply]
I've done this with my custom mediacms.  It's almost as simple as it sounds.  The main issue is ensuring the file copy into the directory is fully complete before processing it.  It sounds easy, but large deployments using network shares/storage might have latency issues or completely lack file change notification abilities.

In practice at my deployment, I basically had several failed attempts per file until I enforced a md5sum match in a secondary file before processing.  It's fairly expensive and can be fairly network intense for shared file systems.

Even with a Cron job or other mechanism, you want to make sure it's present, complete, not locked, etc.

There is also the question of who owns the media?  Admin?  Other?

**Comment by srvprivate** [comment]
It’s a worthy cause for a CMS. I’ve used countless CMS’s over the years in my line of work (Video production) and they all offer that ability. The thing that stands out with yours, is that it is a YouTube replacement, but the code hasn’t branched out beyond that scope. It’s possible, just wondering if it’s a priority. I believe it’s necessary for a lot of people moving forward today.

**Reply by mrjoshuap** [reply]
I think the short version is that it's a small set of devs and between the uploader and API, it seems we have the core functionality in place.

I'm not disagreeing, just starting to think of the requirements.  As I mentioned, I already have this working mostly as you describe.  I do know that my implementation only hits my needs and I didn't consider making it a bit more robust (read: secure, configurable, and generally applicable).

The first challenge is who owns the uploaded media?  The API/UI solves that problem because you have to login.

**Comment by gitwittidbit** [comment]
I just installed MediaCMS because I need something that can (also) act as a frontend to my file server without changing the files themselves and I asked Perplexity which software could this, citing Peertube as an example that can't do this. But now I am coming to realise that MediaCMS also can't do this. 

So, +1 from me for this feature.

And, in the meantime, what are you guys using instead? 

Thanks!

**Comment by mrjoshuap** [comment]
I made a custom script that scans a directory structure and uploads them (yes, I do transcoding) before removing them.

I could theorize, but would doubt it would be a feature making it in... I'm just a dude... You'd effectively need to disable transcoding, rework the web folders so they'd be served, and a script to generate SQL statements to insert the data for the media (depending on your structure, you'd likely need to url encode the paths properly...

Alternatively, you could possibly add a new API method that could take more than the file, description and name...

**Comment by regiesonodl-ui** [comment]
Hello to everyone,

i have a mediaCMS docker installation and it is running very well. 
I need to use the API to automatically upload videos when they are copied in a specified folder on my local machine.
I am on macos here. I understand I will have to go for a watchfolder action and POST to the CMS.
I am a newbie on this and i would like to know if someone could help me on the method to achieve this.
I already have access to the API/swagger and could test the login and even get my user token. 
I would now need some help for the script, either shell script bash or python to monitor the folder and push those videos to the CMS.

Another question: I am not sure if this is possible but I would like those files to be in a category (because RBAC), downloadable and with status private.

Thanks a lot for your help

**Comment by regiesonodl-ui** [comment]
Hello,
I finally managed to do what I asked for in the previous post.
I have a python script that pushes the files through API and it works very well.
I can Upload, set the video private but unfortunately it is yet impossible to choose a category for the files.
I still have to log in later to the cms to assign each file to a category, which makes useless my attempt to automate the process.

Is there a way to implement this category option to the API? And how difficult would it be??

Thanks

**Reply by mgogoulos** [reply]
yes, there's the bulk_actions API (introduced to serve the 'my media' list of actions), also documented in /swagger


Example:

```
POST to /api/v1/media/user/bulk_actions

{"action":"add_to_category","media_ids":["T5NejdTJ4","AQ9Kj1rYq"],"category_uids":["6a376886-4fdb-4d68-a4f3-b2c978fa6b08"]}
```
